lsp/src/tree_sitter_querier.rs

```diff
@@ -0,0 +1,146 @@
+// See: https://tree-sitter.github.io/tree-sitter/using-parsers#query-syntax
+
+use std::collections::HashMap;
+
+use tree_sitter::{Node, Point, Query, QueryCursor, Range};
+
+use crate::tree_sitter::Position;
+
+// If error char is "=" means the key name is completed and the cursor is
+// at the "=" but no quote, so we shouldn't suggest yet eg <div hx-foo=|>
+const KEY_VALUE_SEPARATOR: &str = "=";
+
+fn query_props(
+    query_string: &str,
+    node: Node<'_>,
+    source: &str,
+) -> Option<HashMap<String, String>> {
+    let query = Query::new(tree_sitter_html::language(), query_string).expect(&format!(
+        "get_position_by_query invalid query {query_string}"
+    ));
+    let mut cursor_qry = QueryCursor::new();
+
+    let capture_names = query.capture_names();
+
+    let mut matches = cursor_qry.matches(&query, node, source.as_bytes());
+
+    let mut props = HashMap::new();
+    matches.into_iter().for_each(|match_| {
+        match_.captures.iter().for_each(|capture| {
+            let name = capture_names[capture.index as usize].to_owned();
+            let value = capture
+                .node
+                .utf8_text(source.as_bytes())
+                .expect(&format!("failed to parse capture value for '{name}'"))
+                .to_owned();
+            props.insert(name, value);
+        });
+    });
+
+    Some(props)
+}
+
+pub fn query_attr_keys_for_completion(node: Node<'_>, source: &str) -> Option<Position> {
```

Comment:
I moved the logic of querying for attribute key and name suggestions to their own function because it was getting a bit too complex and long to manage :)